llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -2840,22 +2926,34 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpAccessChain>(SPIRVValue *
     }
 
     // Do the final index if we have one.
-    Type *newBaseEltType = newBase->getType()->getScalarType()->getPointerElementType();
-    Value *resultValue = nullptr;
+    baseEltType = newBaseType;
+    // TODO: This will be removed later.
+    assert(baseEltType == newBase->getType()->getScalarType()->getNonOpaquePointerElementType());
     if (spvAccessChain->isInBounds()) {
-      resultValue = getBuilder()->CreateInBoundsGEP(newBaseEltType, newBase, indices);
+      resultValue = getBuilder()->CreateInBoundsGEP(baseEltType, newBase, indices);
     } else {
-      resultValue = getBuilder()->CreateGEP(newBaseEltType, newBase, indices);
+      resultValue = getBuilder()->CreateGEP(baseEltType, newBase, indices);
     }
 
-    return resultValue;
   } else {
-    Type *const baseEltType = base->getType()->getScalarType()->getPointerElementType();
+    baseEltType = basePointeeType;
+
+    // TODO: This will be removed later
+    assert(baseEltType == base->getType()->getScalarType()->getNonOpaquePointerElementType());
+
     if (spvAccessChain->isInBounds())
-      return getBuilder()->CreateInBoundsGEP(baseEltType, base, indices);
+      resultValue = getBuilder()->CreateInBoundsGEP(baseEltType, base, indices);
     else
-      return getBuilder()->CreateGEP(baseEltType, base, indices);
+      resultValue = getBuilder()->CreateGEP(baseEltType, base, indices);
   }
+
+  // At this point we have to store correlation between SPIR-V AccessChain (SPIRVValue) and its dereferenced return type
+  // (llvm type). It is not possible to get this information later using transType(spvAccessChain->getType()) because if
+  // access chain is pointing to the structure of vetors/matrix then return type may contain some additional padding
+  // which will not be reflected when doing transType.
+  addAccessChainRetType(spvAccessChain, GetElementPtrInst::getIndexedType(baseEltType, indices));
```

Comment:
Just something to keep in mind when you are working on other patches: this may need generalisation as I cannot convince myself that we only need to care about AccessChain as opposed to all instructions that return a pointer.